### PR TITLE
feat: fork ts support

### PR DESCRIPTION
**Pull request by ycjcl868**
image

##### Checklist

- [ ] `npm test` passes
- [ ] tests are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

##### Description of change

- any feature?
- close https://github.com/umijs/umi/ISSUE_URL

### CODE REVIEW

**Comment by ycjcl868** [comment]
这样插件只做类型提示，不做类型检查

File: packages/preset-built-in/src/plugins/registerMethods.ts [hunk 1]
```diff
@@ -53,14 +54,26 @@ export default function (api: IApi) {
...
+      tsNoCheck?: boolean;
+    }) {
       assert(
         api.stage >= api.ServiceStage.pluginReady,
         `api.writeTmpFile() should not execute in register stage.`,
       );
       const absPath = join(api.paths.absTmpPath!, path);
       api.utils.mkdirp.sync(dirname(absPath));
       if (!existsSync(absPath) || readFileSync(absPath, 'utf-8') !== content) {
+        if (/\.(ts|tsx)$/.test(path) && tsNoCheck) {
```
Comment (sorrycc) [inline packages/preset-built-in/src/plugins/registerMethods.ts:73]: .d.ts 应该不加。

File: packages/preset-built-in/src/fixtures/api-writeTmpFile-ts-check/plugin.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (afc163) [inline packages/preset-built-in/src/fixtures/api-writeTmpFile-ts-check/plugin.js:1]: 多了个空行。

File: packages/preset-built-in/src/plugins/registerMethods.ts (hunk 1 above)
Comment (ycjcl868) [reply packages/preset-built-in/src/plugins/registerMethods.ts:73]: ok

File: docs/plugins/api.md [hunk 3]
```diff
@@ -493,14 +493,15 @@ dev 退出时触发。
...
-### writeTmpFile({ path: string, content: string })
+### writeTmpFile({ path: string, content: string, tsNoCheck: boolean })
 
 写临时文件。
 
 参数：
 
 - `path`：相对于临时文件夹的路径
 - `content`：文件内容
+- `tsNoCheck`：默认为 `true`，`path` 为 ts 或 tsx 文件，不检查 TypeScript 类型错误，如果希望插件对用户项目进行 TypeScript 类型检查，可以设置为 `false`。
```
Comment (afc163) [inline docs/plugins/api.md:504]: `skipTsCheck` 吧

File: packages/types/index.d.ts [hunk 4]
```diff
@@ -110,7 +110,7 @@ export interface IApi extends PluginAPI {
   stage: typeof Service.prototype.stage;
   ServiceStage: typeof Service.prototype.ServiceStage;
   writeTmpFile: {
-    (args: { path: string; content: string; tsNoCheck?: boolean }): void;
+    (args: { path: string; content: string; skipTSCheck?: boolean }): void;
```
Comment (afc163) [inline packages/types/index.d.ts:113]: `skipTSCheck` or `skipTsCheck`?

Comment (ycjcl868) [reply packages/types/index.d.ts:113]: 改成 `skipTsCheck` 吧， 我再提个 PR

Comment (ycjcl868) [reply packages/types/index.d.ts:113]: 看配置用的 `forkTSChecker`，https://umijs.org/config#forktschecker，统一成 `TSCheck`？

**Pull request merged**